src/NHibernate.Test/Async/NHSpecificTest/NH2319/Fixture.cs

```diff
@@ -0,0 +1,428 @@
+				AssertFilterPlanCacheHit(spy);
+			}
+		}
+
+		[Test]
+		public async Task ShouldNotInitializeCollectionWhenPerformingQueryAsync()
+		{
+			using (var session = OpenSession())
+			using (session.BeginTransaction())
+			{
+				var parent = await (session.GetAsync<Parent>(_parent1Id));
+				Assert.That(parent, Is.Not.Null);
+
+				var persistentCollection = (IPersistentCollection) parent.Children;
+
+				var filtered = await (parent.Children
+					.AsQueryable()
+					.Where(x => x.Name == "Jack")
+					.ToListAsync());
+
+				Assert.That(filtered, Has.Count.EqualTo(1));
+				Assert.That(persistentCollection.WasInitialized, Is.False);
+			}
+		}
+
+		[Test]
+		public async Task ShouldPerformSqlQueryEvenIfCollectionAlreadyInitializedAsync()
+		{
+			using (var session = OpenSession())
+			using (session.BeginTransaction())
+			{
+				var parent = await (session.GetAsync<Parent>(_parent1Id));
+				Assert.That(parent, Is.Not.Null);
+
+				var loaded = parent.Children.ToList();
+				Assert.That(loaded, Has.Count.EqualTo(2));
+
+				var countBeforeFiltering = session.SessionFactory.Statistics.QueryExecutionCount;
+
+				var filtered = await (parent.Children
+					.AsQueryable()
+					.Where(x => x.Name == "Jack")
+					.ToListAsync());
+
+				var countAfterFiltering = session.SessionFactory.Statistics.QueryExecutionCount;
+
+				Assert.That(filtered, Has.Count.EqualTo(1));
+				Assert.That(countAfterFiltering, Is.EqualTo(countBeforeFiltering + 1));
+			}
+		}
+
+		[Test]
+		public async Task TestFilterAsync()
+		{
+			using (var session = OpenSession())
+			using (session.BeginTransaction())
+			{
+				var parent = await (session.GetAsync<Parent>(_parent1Id));
+				Assert.That(parent, Is.Not.Null);
+
+				var children = await ((await (session.CreateFilterAsync(parent.Children, "where this.Name = 'Jack'")))
+					.ListAsync<Child>());
+
+				Assert.That(children, Has.Count.EqualTo(1));
+			}
+		}
+
+		[Test]
+		public async Task TestPlanCacheMissAsync()
+		{
+			var internalPlanCache = typeof(QueryPlanCache)
+				.GetField("planCache", BindingFlags.NonPublic | BindingFlags.Instance)
+				?.GetValue(Sfi.QueryPlanCache) as SoftLimitMRUCache;
+			Assert.That(internalPlanCache, Is.Not.Null,
+				$"Unable to find the internal query plan cache for clearing it, please adapt code to current {nameof(QueryPlanCache)} implementation.");
+
+			using (var spy = new LogSpy(typeof(QueryPlanCache)))
+			{
+				internalPlanCache.Clear();
+				await (ShouldBeAbleToFindChildrenByNameAsync());
+				AssertFilterPlanCacheMiss(spy);
+			}
+		}
+
+		private static void AssertFilterPlanCacheHit(LogSpy spy) =>
+			// Each query currently ask the cache two times, so asserting reuse requires to check cache has not been missed
+			// rather than only asserting it has been hit.
+			Assert.That(spy.GetWholeLog(),
+				Contains.Substring("located collection-filter query plan in cache (")
+				.And.Not.Contains(_filterPlanCacheMissLog));
```

Comment:
[#49](https://github.com/maca88/AsyncGenerator/issues/49), missing const in async file.